### ISSUE TITLE
New Task: Update deployment start/end dates in portal when updated in ADR

### ISSUE THREAD

**Issue by iwensu0313** [ioos/marine_life_data_network#110]
### Who is requesting this?

ATN DAC original funders, manufacturers, & data contributors

### What is being requested?

Deployment start and end dates from the hardware by default include times when the equipment is out of water. We need to fix the deployment visualizations in the portal so that they only show the data and location tracking for the periods when the equipment is actually in the water.

### What is the current status quo (i.e., what happens if this does not get done)?

If this is not done, users will continue to see data and tracks before and after the equipment is actually deployed in the water, which can cause confusion.

### What indicates this is done (i.e., how do we know this is complete)?

Deployment visualizations accurate align with the start/end dates manually inputted in the ATN DAC Registration portal, which should reflect the actual dates of deployment.

- [ ] Enable ATN DAC Registration (ADR) to speak with Axiom data ingestion system
- [ ] Implement time slicing in Axiom data ingestion system based on ADR most up-to-date start/end deployment metadata?
- [ ] Re-run deployments after system changes have been implemented and tested?

NOTE: I need to update the specific steps in this section after confirming with our engineers that this is still the best proposed pathway.

**Comment by iwensu0313** [comment]
Nov update from team
- Update: In staging environment, demonstrated updating a deployment in ADR and having it trigger our ingestion pipeline so that already visualized deployments in the portal are updated with new information (for clarity, the PIs need to input the updated truncation dates in ADR for this to be updated in the portal)
- In Progress: working on some staging fixes/tests before we can deploy to production
- Note: this does not address issues that arise when the deployment id is not provided in the ADR

Hoping to wrap up by end of month, else will need to pause Dec 1.

**Comment by iwensu0313** [comment]
Dec update: still need a few more days of work to bring changes to production. Paused Dec 1 for now, until prioritized with IOOS Marine Life/ATN team.

**Comment by iwensu0313** [comment]
Discussed at ATN data coord meeting today that this is a big issue for data providers. Results in messiness of the visualization. We decided to prioritize since we are almost at the finish line. (cc @mmckinzie , @MathewBiddle @conniekot). Moved to IN PROGRESS.

**Comment by laurabrenskelle** [comment]
@iwensu0313 Are there any updates on this task?

**Comment by iwensu0313** [comment]
Update: initial release of ADR-data ingestion pipeline connection development has been completed. We are currently in a testing phase, which is required to ensure that improvements are meeting the needs of the project. Here are some possible next steps

- Verification Testing of Production (quick): designate a project that Axiom staff can test in ADR (saving a small metadata change, can revert) to ensure and observe that the behavior of improvements in production is the same as the behavior in staging (highly recommended, low effort)
- Test that we can update Start/End Dates for already ingested projects with the new improvements: truncation of prior ingested deployments/projects will need to be manually re-run; for a subset of all ATN deployments, this will be straightforward since they already have the required metadata (recommended, medium effort)

Additional effort has been paused as of Feb 27, 2025 until further discussion.

**Comment by MathewBiddle** [comment]
@iwensu0313 if able "Verification testing of production" is okay to spend some effort on.  @mmckinzie can you support here?

**Comment by mmckinzie** [comment]
@iwensu0313 

If you plan to reprocess an entire project to verify that data are now being truncated as expected in the portal based on reported metadata in ADR, these two projects would be great candidates for testing, as they provide ample opportunities to verify success or document any lingering issues or an unexpected response. 

Hammerschlag: [Satellite Tracking of Migratory Shark Species, 2010-2023](https://dacregistration.atn.ioos.us/view/84/) (ADR project # 84) 
Southall: [Atlantic Behavioral Response Study, 2017-2024](https://dacregistration.atn.ioos.us/view/26/) (ADR project # 26)

If you're looking to manipulate metadata in the ADR to test visualization response in the portal, then this is a clean project with more limited deployments that have nice, clear tracks.

Huckstadt: [NSF-NERC Collaborative Research: Effects of a Changing Climate on the Distribution, Abundance and Habitat Utilization of Crabeater Seals, 2022-2024](https://dacregistration.atn.ioos.us/view/168/) (ADR project # 168)

**Comment by iwensu0313** [comment]
Thanks @mmckinzie - we did a quick test last week and here is the update:

- ADR connection with our data ingestion pipeline is successful (non trivial effort - I don't know of many other places where external users can affect/trigger an internal Axiom pipeline, so we had to tread carefully and is a big win)
- data ingestion pipeline is failing to carry updated metadata to portal (would require additional troubleshooting of the code to ensure - PAUSING UNLESS WE DECIDE TO PRIORITIZE)

**Comment by iwensu0313** [comment]
@MathewBiddle okay to pick back up in the new year? Let me know if you want to discuss first. 

As a reminder, the goal is to carry updated metadata (particularly start/end dates) provided by the data provider in the ATN Data Registration portal all the way through downstream where it is reflected in the ATN Portal. We finished pushing through on the backend changes earlier this year and entered the testing phase. We were able to do an initial test and noted that the first part of the update works as expected, but the second half needs to be looked into. That's when we paused. 

On a possible related note, we have a single deployment that recently didn't pass the NCEI archive pipeline because it had lat/lons on land. And the seanames package didn't know how to address that. I wanted to see if this effort could resolve that edge case as well (e.g. in cases where the land lat/lons are not supposed to be included in the archive) or if it's case where sea names needs to ignore land lat/lons. Ideally the former so that we're passing correct intended data to the sea names pkg.

**Comment by MathewBiddle** [comment]
@iwensu0313 go forth and conquer. Keep us posted on what the level of effort will be to fix this.

**Comment by iwensu0313** [comment]
For real-time data ingests (those that pull data directly from WC or SMRU APIs/servers), updates to ADR metadata should automatically trigger updates downstream in the corresponding metadata in the ATN Portal.

For manual ingests (source is RW uploads), updates to ADR metadata still need to be manually triggered for those metadata to update in the ATN Portal. More investigation will have to be done to determine feasibility

Example application: PIs often/not infrequntly update the start/end date in ADR to truncate auto-ingests to properly represent the actual in water deployment days. This should update in the ATN portal now so that data does not appear on land and cause wonky trajectories that might obscure users from being able to interpret the data. 

@mmckinzie  @MathewBiddle - marking as DONE. 

There are improved error logs so that if the pipeline runs into issues w/ duplicate deployment IDs in our system and are unable to identify which one is the correct one, there should be info logs so that we can manually update the date truncation. Though first assumption is to grab the PTT ID from WC API to identify the correct deployment if pipeline runs into this edge case.
